### PR TITLE
add constant learning rate with custom rule

### PR DESCRIPTION
**Pull request by jason9075**
You can write rule to control constant lr

Ex:
lr="0.0005"
rules="1:10,0.1:20,0.01:30,0.005"

it means that the learning rate is multiple 1 for the first 10 steps, mutiple 0.1 for the next 20 steps, multiple 0.01 for the next 30 steps and multiple 0.005 for the other steps.

lr = 0.0005 from step 0 ~ 9
lr = 0.00005 from step 10 ~ 19
lr = 0.000005 from step 20 ~ 29
lr = 0.0000025 after step 30

### CODE REVIEW

**Comment by pcuenca** [comment]
I've seen this sort of schedule in the past, for long training runs where it's desired to periodically reduce the lr when loss plateaus. Personally, I just kill the training run and resume from a saved checkpoint if I want to decrease the lr, but this approach could be interesting for reproduction purposes, if there are researchers that publish such an schedule in their paper. I'm unaware about any such occurrences in the diffusion area.

Maybe @jason9075 could clarify the use-case they are trying to resolve with this addition. I'd be really happy to learn about the motivation :)

**Comment by jason9075** [comment]
@pcuenca 
When I am fine-tuning the model, I will replace the learning rate of cosine with a constant to reduce the loss value.
For example: I have cosine lr(magenta color line) to get a good model.
Screenshot from 2023-04-26 22-17-04

 And now I can use constant with rules(yellow line) to adjust lr to reduce loss more.
Screenshot from 2023-04-26 22-17-30

File: src/diffusers/optimization.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ class SchedulerType(Enum):
     POLYNOMIAL = "polynomial"
     CONSTANT = "constant"
     CONSTANT_WITH_WARMUP = "constant_with_warmup"
+    CONSTANT_WITH_RULES = "constant_with_rules"
```
Comment (pcuenca) [inline src/diffusers/optimization.py:37]: I might not be familiar enough with this scheduler, but "constant with rules" does not mean a lot to me. Could we use something like "piecewise constant" or "constant with steps"?

File: src/diffusers/optimization.py [hunk 2]
```diff
@@ -77,6 +78,48 @@ def lr_lambda(current_step: int):
     return LambdaLR(optimizer, lr_lambda, last_epoch=last_epoch)
 
 
+def get_constant_schedule_with_rules(optimizer: Optimizer, rules: str, last_epoch: int = -1):
```
Comment (pcuenca) [inline src/diffusers/optimization.py:81]:
```suggestion
def get_piecewise_constant_schedule(optimizer: Optimizer, steps: str, last_epoch: int = -1):
```

If we agree on the naming suggested above

File: src/diffusers/optimization.py [hunk 3]
```diff
@@ -77,6 +78,48 @@ def lr_lambda(current_step: int):
...
+    """
+    Create a schedule with a constant learning rate with rule for the learning rate.
+
+    Args:
+        optimizer ([`~torch.optim.Optimizer`]):
+            The optimizer for which to schedule the learning rate.
+        rule (`string`):
```
Comment (pcuenca) [inline src/diffusers/optimization.py:88]: Maybe rename to `steps` or `step_rules`

File: src/diffusers/optimization.py [hunk 4]
```diff
@@ -77,6 +78,48 @@ def lr_lambda(current_step: int):
...
+        def rule_func(steps: int) -> float:
+            sorted_steps = sorted(rules_dict.keys())
+            for i, sorted_step in enumerate(sorted_steps):
+                if steps < sorted_step:
+                    return rules_dict[sorted_steps[i]]
+            return last_lr
+
+        return rule_func
+
+    rules_f = create_rules_function(rules_dict, last_lr)
```
Comment (pcuenca) [inline src/diffusers/optimization.py:118]:
```suggestion
    last_lr_multiple = float(rule_list[-1])

    def create_rules_function(rules_dict, last_lr_multiple):
        def rule_func(steps: int) -> float:
            sorted_steps = sorted(rules_dict.keys())
            for i, sorted_step in enumerate(sorted_steps):
                if steps < sorted_step:
                    return rules_dict[sorted_steps[i]]
            return last_lr_multiple

        return rule_func

    rules_f = create_rules_function(rules_dict, last_lr_multiple)
```

File: src/diffusers/optimization.py [hunk 5]
```diff
@@ -232,12 +275,14 @@ def lr_lambda(current_step: int):
...
     SchedulerType.CONSTANT: get_constant_schedule,
     SchedulerType.CONSTANT_WITH_WARMUP: get_constant_schedule_with_warmup,
+    SchedulerType.CONSTANT_WITH_RULES: get_constant_schedule_with_rules,
 }
 
 
 def get_scheduler(
     name: Union[str, SchedulerType],
     optimizer: Optimizer,
+    rules: Optional[str] = None,
```
Comment (pcuenca) [inline src/diffusers/optimization.py:285]: Can we also add to the docstring explaining that it only applies to the new schedule?

**Review by pcuenca — COMMENTED** [review]
Looks good to me, just suggested some naming changes trying to achieve more clarity.

File: src/diffusers/optimization.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/optimization.py:37]: Piecewise constants sounds better and in line with what I have seen being used in the other libraries too.

**Comment by sayakpaul** [comment]
Nothing to add on top of what @pcuenca mentioned already. 

Btw, I would appreciate some further clarifications on https://github.com/huggingface/diffusers/pull/3133#issuecomment-1523517527. 

> For example: I have cosine lr(magenta color line) to get a good model.

Are you referring to the initial model which you'd later fine-tune? 

> And now I can use constant with rules(yellow line) to adjust lr to reduce loss more.


The figure below this statement -- have you zoomed in on the losses from the two different setups? To me, they don't seem to be very different, though. Could you help me understand this bit?

File: src/diffusers/optimization.py (hunk 1 above)
Comment (jason9075) [reply src/diffusers/optimization.py]: You are right. I should change the name to piecewise constant.

**Comment by jason9075** [comment]
> Nothing to add on top of what @pcuenca mentioned already.
> 
> Btw, I would appreciate some further clarifications on [#3133 (comment)](https://github.com/huggingface/diffusers/pull/3133#issuecomment-1523517527).
> 
> > For example: I have cosine lr(magenta color line) to get a good model.
> 
> Are you referring to the initial model which you'd later fine-tune?
> 
> > And now I can use constant with rules(yellow line) to adjust lr to reduce loss more.
> 
> The figure below this statement -- have you zoomed in on the losses from the two different setups? To me, they don't seem to be very different, though. Could you help me understand this bit?

I just remove the log. So I retrain another one with 3000 steps.

Screenshot from 2023-04-27 21-52-26
I finetune the same model with same dataset but with different scheduler separately. Most of time cosine is good enough to get good result. But I will do some experiment to adjust different learning rate. 

Here is zoom in of loss of graph. I think it may be that I used very few parameters(LoRA), which caused the loss graph to be very similar.
Screenshot from 2023-04-27 21-52-58

**Comment by sayakpaul** [comment]
Fair enough. Would be good to have some results with non-LoRA as well. Do you think that would be possible?

**Comment by patrickvonplaten** [comment]
Ok to merge this from my point of view

**Comment by sayakpaul** [comment]
Yeah my questions are non-blocker. Sorry for not making that 100% clear. Merging.
